alt_e2eshark/e2e_testing/backends.py

```diff
@@ -0,0 +1,41 @@
+# This file will contain customizations for how to compile mlir from various entrypoints
+import abc
+from typing import TypeVar
+
+CompiledArtifact = TypeVar("CompiledArtifact")
+Invoker = TypeVar("Invoker")
+
+class BackendBase(abc.ABC):
+
+    @abc.abstractmethod
+    def compile(self, module) -> CompiledArtifact:
+        '''specifies how to compile an MLIR Module'''
+
+    @abc.abstractmethod
+    def load(self, artifact: CompiledArtifact) -> Invoker:
+        '''loads the compiled artifact'''
+
+
+from iree import compiler as ic
+from iree import runtime as rt
+from torch_mlir.passmanager import PassManager
+
+class SimpleIREEBackend(BackendBase):
+
+    def compile(self, module):
+        pipeline = "builtin.module(func.func(convert-torch-onnx-to-torch))"
+        with module.context as ctx:
+            pm = PassManager.parse(pipeline)
+            pm.run(module.operation)
+        return ic.tools.compile_str(str(module), input_type="torch", target_backends=["llvm-cpu"])
```

Comment:
Yeah, good point. Would it be easy to add support for input_type="torch-onnx" at this point, for example? Or is it difficult for the AUTO option to detect the difference between torch-onnx and torch since they are using the same dialect?

Reply:
Here is the current logic in IREE:
* https://github.com/iree-org/iree/blob/34449df33fbcc37d680abf37fdcfeae875ac04a1/compiler/plugins/input/Torch/PluginRegistration.cpp#L119-L125
    ```c++
        // ONNX is considered a superset of Torch. It runs all of the Torch
        // pipelines with an extra ONNX-specific preprocessing step.
        if (hasOnnx) {
          typeMnemonics.insert("onnx");
        } else if (hasTorch) {
          typeMnemonics.insert("torch");
        }
    ```
* https://github.com/iree-org/iree/blob/34449df33fbcc37d680abf37fdcfeae875ac04a1/compiler/plugins/input/Torch/PluginRegistration.cpp#L58-L69
    ```c++
        if (typeMnemonic == "onnx") {
          // ONNX input is a pre-processing step to torch.
          passManager.addNestedPass<func::FuncOp>(
              mlir::torch::onnx_c::createTorchOnnxToTorchPass());
        }
    
        if (typeMnemonic == "torch" || typeMnemonic == "onnx") {
          TorchInput::TorchToIREELoweringPipelineOptions torchOptions;
          torchOptions.strictSymbolicShapes = options.strictSymbolicShapes;
          TorchInput::createTorchToIREEPipeline(passManager, torchOptions);
          return true;
        }
    ```

We could change that to always run `createTorchOnnxToTorchPass` if any torch ops are found, or just keep the current behavior.